Various non-coherent memory handling fixes

mark-lunarg opened the pull request:
Hopefully for #819 , #820 , and #822 , along with a test case that exposed the crashes on NCM systems.

mark-lunarg commented:
Gary, if you can, please test these on your setup there.  I hacked up the layers to do some testing but we won't have an actual non-coherent memory system until later this week.

tobine commented on layers/core_validation.cpp:
```diff
@@ -5008,10 +5008,15 @@ static void initializeAndTrackMemory(layer_data *dev_data, VkDeviceMemory mem, V
+            // Ensure start of mapped region reflects hardware alignment constraints
+            size_t map_alignment = dev_data->phys_dev_properties.properties.limits.minMemoryMapAlignment;
+            mem_info->p_data = malloc(2 * map_size + map_alignment);
+            mem_info->alignment_offset = reinterpret_cast<uintptr_t>(mem_info->p_data) % map_alignment;
```
It's a bit clunky to carry the alignment_offset around to continually increment p_data and I'm afraid it will be error-prone.

I think the only place we'll need the original p_data after allocation is when the memory is freed in deleteMemRanges().
One alternative is to increment p_data from the start (as soon as alignment_offset is calculated). That then avoids all of the future additions of alignment_offset, but instead when the memory is freed p_data will need to be decremented by alignment_offset. That's still not great but I'd prefer that over this current code.

Another option would be to not store the alignment offset, but instead store "p_data" (allocation address plus alignment offset) and "p_data_base" (allocation address). Then p_data is used everywhere except when memory is freed and "p_data_base" would be used. I don't have a preference for this option, just putting it out as another alternative.

mark-lunarg replied:
I prefer the option of saving the pointer and the base address.  I'll do the rework, but Gary can do the testing with what he has, if he is able.

tobine commented on tests/layer_validation_tests.cpp:
```diff
@@ -2181,6 +2182,21 @@ TEST_F(VkLayerTest, InvalidMemoryMapping) {
     vkFlushMappedMemoryRanges(m_device->device(), 1, &mmr);
     m_errorMonitor->VerifyFound();
 
+    // Map using WHOLE_SIZE, then Flush and Invalidate using same.
```
I'm not a fan of including this expected passing test as part of "InvalidMemoryMapping" test.
Should be broken out into a separate test (best option IMO) or the test should be renamed to account for increased scope.

tobine commented:
With those couple clean-ups I noted 
LGTM :+1:

gary-sweet commented:
Ok - I've given this a test now.

#819 looks fixed.
#822 looks fixed.

#820 has a bug:

You have aligned the size of the start of the memory block, but you are returning that plus the size of the front guard band (size/2), which then misaligns what you are returning again. You need to make sure the address returned is aligned, not the start of the guard banded block.

If you add "assert(reinterpret_cast<uintptr_t>(*ppData) & (map_alignment - 1) == 0);" at the end of the function, this would trap any mistakes in the arithmetic.

mark-lunarg commented:
Yeah, we had seen that yesterday and I added a fix along with the changes for Tobin's review suggestions.  I'll add the assert as well and have something for you to try shortly.

mark-lunarg commented:
I reworked 822 pretty severely, changing the pad size from half the allocation size to just the granularity size, added the assert as Gary suggested.  Please have another look at this commit if you get a chance.

tobine commented on layers/core_validation.cpp:
```diff
@@ -5003,18 +5005,30 @@ static void initializeAndTrackMemory(layer_data *dev_data, VkDeviceMemory mem, V
+            mem_info->shadow_copy_base = 0;
+            mem_info->shadow_copy = 0;
         } else {
+            uint64_t shadow_pad_size = dev_data->phys_dev_properties.properties.limits.minMemoryMapAlignment;
```
Confusing that you declare "shadow_pad_size" up here apart from map_alignment below. It appears that this var is more for convenience as it has same value as map_alignment. Would be nice to declare it below with a comment. If it's truly the same as map_alignment throughout it could either be eliminated or assigned directly from map_alignment.

mark-lunarg replied:
I purposely decoupled them so that we could easily change the pad size, so I'd like to keep them separate but will move the line and add comment.

tobine commented on tests/layer_validation_tests.cpp:
```diff
@@ -2197,6 +2197,72 @@ TEST_F(VkLayerTest, InvalidMemoryMapping) {
+TEST_F(VkLayerTest, NonCoherentMemoryMapping) {
+
+    TEST_DESCRIPTION("Ensure that validations handling of non-coherent memory "
+                     "mapping while using VK_WHILE_SIZE does not cause access "
```
WHILE -> WHOLE

gary-sweet commented:
Almost, but not quite I'm afraid.

If I call vkMapMemory with an offset that is non-zero, I get the start address of the shadow block back. The offset has been ignored.

mark-lunarg commented:
OK, updated for feedback and hopefully addressed the ignored offsets.  Gary, can you please give this another shot?  Sorry to keep asking, but our Android build guy has been difficult to pin down lately.

gary-sweet commented:
This is passing all of my tests here now.
Thanks

mark-lunarg commented:
Thanks again, and keep the issues coming.  Pushed to Master.

commit 5f06b9eba1bf89d4066ca44120dcb72c4cf44104
Author: Mark Lobodzinski <mark@lunarg.com>
Date:   Mon Aug 15 14:27:26 2016 -0600

    layers: GH820, MapMemory shadow copy alignment

    For host-visible non-coherent memory, the validation layers use a
    shadow copy of the mapped memory region. Added proper alignment
    for the shadow copy, fixed handling of offsets, and implemented
    resolve of the shadow data on invalidate calls.

    Change-Id: I88815799865d3a30c76e16cf96ace630e1bbcf2d
